File: fastq_rnd_seek_proper.jl
#!/Applications/Julia-0.3.8.app/Contents/Resources/julia/bin/julia
# required modules
require("argparse")
using ArgParse

# function to parse arguments
function parse_arguments()
  s = ArgParseSettings()
  @add_arg_table s begin
    "filename"
      help = "Input filename."
      required = true
    "numsamples"
      help = "Number of sequences to sample."
      arg_type = Int
      default = 10000
  end
  return parse_args(s)
end

# function to return random position in file
function get_random_pos(rnd_num)
  return rand(0:rnd_num)
end

# main function
function main()
  # retrieve parsed arguments
  parsed_args = parse_arguments()

  # set filename
  filename = parsed_args["filename"]
  numsamples = parsed_args["numsamples"]

  # check file exists
  if !isfile(filename)
    throw(LoadError("", 0, "Filename $(filename) not found."))
  end

  # get file size
  infilesize = filesize(filename)

  # open input file
  infh = open(filename, "r")

  # iterate number of samples times
  for i = 1:numsamples
    valid_record = false
    lines = Array(ASCIIString, 4)
    while !valid_record
      pos = get_random_pos(infilesize)
      seek(infh, pos)
      readuntil(infh, "\n@")
      pos = position(infh) - 1
      seek(infh, pos)
      for j in [1:4]
        lines[j] = rstrip(readline(infh))
      end
      if beginswith(lines[1], "@") && beginswith(lines[3], "+")
        valid_record = true
      end
    end
    println(join(lines, "\n"))
  end

  # close file handle
  close(infh)
end

# call main function
main()

File: fastq_rnd_seek_proper.py
#!/usr/bin/env python
# import modules
import mmap, os, random, sys
from Bio import SeqIO
from StringIO import StringIO

# parse arguments
if len(sys.argv) != 3:
  print "Requires an input filename and integer as arguments."
  sys.exit()
filename = sys.argv[1]
try:
  num_seqs = int(sys.argv[2])
except:
  print "Require a valid integer value for number of sequences to sample."
  sys.exit()

# check file exists
if not os.path.exists(filename):
  print "Filename %s doesn't exist." % filename
  sys.exit()

# get size of the file
filesize = os.path.getsize(filename)

# define variables
observed_offsets = []

# open file
with open(filename, 'rb') as infh:
  mm = mmap.mmap(infh.fileno(), 0, prot=mmap.PROT_READ)
  for i in range(num_seqs):
    failed = True
    while failed:
      # pick a random offset and seek to that position
      random_offset = random.randrange(filesize)
      mm.seek(random_offset)
      # always seek to a position where there is a record in the file
      if mm.rfind("\n@") == -1:
        continue
      # find the first record after the offset
      current_record_offset = mm.find("\n@") + 1
      # check not already seen this record
      if current_record_offset in observed_offsets:
        continue
      # add current record to list of observed records 
      observed_offsets.append(current_record_offset)
      # seek to current record and find the next
      mm.seek(current_record_offset)
      next_record_offset = mm.find("\n@")
      # check not already in the last record set as end of file if we are
      if next_record_offset == -1:
        next_record_offset = filesize
      # read record
      seq = mm.read(next_record_offset - current_record_offset)
      # check if is a valid sequence
      try:
        record = SeqIO.read(StringIO(seq), 'fastq')
      except:
        continue
      failed = False
  mm.close()
